internal/cloudapi/v2/handler.go

```diff
@@ -1227,3 +1227,226 @@ func genRepoConfig(repo Repository) (*rpmmd.RepoConfig, error) {
+		}
+		if len(awsT.ShareWithAccounts) > 0 {
+			shares = append(shares, awsT.ShareWithAccounts...)
+		}
+		if img.ShareWithAccounts != nil && len(*img.ShareWithAccounts) > 0 {
+			shares = append(shares, (*img.ShareWithAccounts)...)
+		}
+		if len(shares) > 0 {
+			shareJob := &worker.AWSEC2ShareJob{
+				Ami:               shareAmi,
+				Region:            shareRegion,
+				ShareWithAccounts: shares,
+			}
+			finalJob, err = h.server.workers.EnqueueAWSEC2ShareJob(shareJob, finalJob, channel)
+			if err != nil {
+				return HTTPErrorWithInternal(ErrorEnqueueingJob, err)
+			}
+		}
+	} else {
+		return HTTPError(ErrorUnsupportedImage)
+	}
+
+	return ctx.JSON(http.StatusCreated, ImageFromComposeResponse{
+		ObjectReference: ObjectReference{
+			Href: fmt.Sprintf("/api/image-builder-composer/v2/composes/%v/image", jobId),
+			Id:   finalJob.String(),
+			Kind: "ImageFromComposeId",
+		},
+		Id: finalJob.String(),
+	})
+}
+
+func (h *apiHandlers) GetImageFromComposeStatus(ctx echo.Context, id string) error {
+	return h.server.EnsureJobChannel(h.getImageFromComposeStatus)(ctx, id)
+}
+
+func (h *apiHandlers) getImageFromComposeStatus(ctx echo.Context, id string) error {
+	jobId, err := uuid.Parse(id)
+	if err != nil {
+		return HTTPError(ErrorInvalidComposeId)
+	}
+
+	jobType, err := h.server.workers.JobType(jobId)
+	if err != nil {
+		return HTTPError(ErrorComposeNotFound)
+	}
+
+	var us UploadStatus
+	switch jobType {
+	case worker.JobTypeAWSEC2Copy:
+		var result worker.AWSEC2CopyJobResult
+		info, err := h.server.workers.AWSEC2CopyJobInfo(jobId, &result)
+		if err != nil {
+			return HTTPError(ErrorGettingAWSEC2JobStatus)
+		}
+
+		us = UploadStatus{
+			Status: uploadStatusFromJobStatus(info.JobStatus, result.JobError),
+			Type:   UploadTypesAws,
+			Options: AWSEC2UploadStatus{
+				Ami:    result.Ami,
+				Region: result.Region,
+			},
+		}
+	case worker.JobTypeAWSEC2Share:
+		var result worker.AWSEC2ShareJobResult
+		info, err := h.server.workers.AWSEC2ShareJobInfo(jobId, &result)
+		if err != nil {
+			return HTTPError(ErrorGettingAWSEC2JobStatus)
+		}
+
+		us = UploadStatus{
+			Status: uploadStatusFromJobStatus(info.JobStatus, result.JobError),
+			Type:   UploadTypesAws,
+			Options: AWSEC2UploadStatus{
+				Ami:    result.Ami,
+				Region: result.Region,
+			},
+		}
+	default:
+		return HTTPError(ErrorInvalidJobType)
+	}
+
+	return ctx.JSON(http.StatusOK, ImageFromComposeStatus{
+		ObjectReference: ObjectReference{
+			Href: fmt.Sprintf("/api/image-builder-composer/v2/images/%v", jobId),
+			Id:   jobId.String(),
+			Kind: "ImageFromComposeStatus",
+		},
+		UploadStatus: us,
+	})
+}
+
+func uploadStatusFromJobStatus(js *worker.JobStatus, je *clienterrors.Error) UploadStatusValue {
+	if je != nil || js.Canceled {
+		return UploadStatusValueFailure
+	}
+
+	if js.Started.IsZero() {
+		return UploadStatusValuePending
+	}
+
+	if js.Finished.IsZero() {
+		return UploadStatusValueRunning
+	}
+	return UploadStatusValueSuccess
+}
```

Comment:
It would be great if this could be used also in https://github.com/osbuild/osbuild-composer/blob/af65baa6fe38364f28c1b6630ea1303c221091c6/internal/cloudapi/v2/handler.go#L644-L645 and https://github.com/osbuild/osbuild-composer/blob/af65baa6fe38364f28c1b6630ea1303c221091c6/internal/cloudapi/v2/handler.go#L701-L702
so that we are determining the upload status consistently in the implementation.